File: anti_matter/app/zwave_payload.py
"""Z-Wave SmartStart QR (SDS13937) — parity with zwave-js/qr."""

from __future__ import annotations

import hashlib
import re
from typing import Any

LEAD_IN = 90
VERSION_SMART_START = 1
# Matches zwave-js's ProvisioningInformationType enum (packages/core/src/qr/definitions.ts).
TLV_PRODUCT_TYPE = 0
TLV_PRODUCT_ID = 1
TLV_SUPPORTED_PROTOCOLS = 4


def _digits_only(value: str) -> str:
    return re.sub(r"\D", "", value or "")


def format_dsk(dsk_digits: str) -> str:
    d = _digits_only(dsk_digits)
    if len(d) != 40:
        return (dsk_digits or "").strip()
    return "-".join(d[i : i + 5] for i in range(0, 40, 5))


def parse_dsk_groups(dsk_digits: str) -> list[int] | None:
    d = _digits_only(dsk_digits)
    if len(d) != 40:
        return None
    groups: list[int] = []
    for i in range(8):
        g = int(d[i * 5 : i * 5 + 5], 10)
        if g < 0 or g > 65535:
            return None
        groups.append(g)
    return groups


def is_valid_dsk_formatted(value: str) -> bool:
    s = (value or "").strip()
    if not re.fullmatch(r"\d{5}(-\d{5}){7}", s):
        return False
    return parse_dsk_groups(s) is not None


def pin_from_dsk(dsk_digits: str) -> str:
    d = _digits_only(dsk_digits)
    return d[:5] if len(d) >= 5 else ""


def checksum_valid(qr_digits: str) -> bool:
    if len(qr_digits) < 9:
        return False
    given = int(qr_digits[4:9])
    body = qr_digits[9:]
    digest = hashlib.sha1(body.encode("ascii")).digest()
    expected = (digest[0] << 8) | digest[1]
    return given == expected


def _parse_tlvs(tail: str) -> dict[str, Any]:
    meta: dict[str, Any] = {}
    pos = 0
    while pos + 4 <= len(tail):
        type_crit = int(tail[pos : pos + 2])
        pos += 2
        typ = type_crit >> 1
        length = int(tail[pos : pos + 2])
        pos += 2
        if length < 0 or pos + length > len(tail):
            break
        data = tail[pos : pos + length]
        pos += length
        if typ == TLV_PRODUCT_TYPE and length >= 10:
            word = int(data[0:5])
            meta["generic_device_class"] = word >> 8
            meta["specific_device_class"] = word & 0xFF
            meta["installer_icon_type"] = int(data[5:10])
        elif typ == TLV_PRODUCT_ID and length >= 20:
            meta["manufacturer_id"] = int(data[0:5])
            meta["product_type"] = int(data[5:10])
            meta["product_id"] = int(data[10:15])
            app = int(data[15:20])
            meta["application_version"] = f"{app >> 8}.{app & 0xFF}"
        elif typ == TLV_SUPPORTED_PROTOCOLS and length in (2, 3, 5):
            bits = int(data)
            meta["supported_protocols"] = {
                "zwave": bool(bits & 0x01),
                "zwave_long_range": bool(bits & 0x02),
            }
    return meta


def _parse_requested_security_classes(requested_keys_digits: str) -> dict[str, bool]:
    """RequestedKeys byte (3 decimal digits at QR position 9-12) — bit N corresponds to
    SecurityClass ordinal N in zwave-js: 0=S2_Unauthenticated, 1=S2_Authenticated,
    2=S2_AccessControl, 7=S0_Legacy (bits 3-6 are unused/reserved)."""
    try:
        bits = int(requested_keys_digits)
    except ValueError:
        bits = 0
    return {
        "s2_unauthenticated": bool(bits & 0x01),
        "s2_authenticated": bool(bits & 0x02),
        "s2_access_control": bool(bits & 0x04),
        "s0_legacy": bool(bits & 0x80),
    }


def parse_qr_digits(qr_digits: str) -> dict[str, Any] | None:
    d = _digits_only(qr_digits)
    if len(d) < 52 or not d.startswith("90"):
        return None
    version = int(d[2:4])
    dsk_raw = d[12:52]
    if parse_dsk_groups(dsk_raw) is None:
        return None
    if not checksum_valid(d):
        return None
    meta = _parse_tlvs(d[52:])
    return {
        "qr": d,
        "dsk": format_dsk(dsk_raw),
        "pin": pin_from_dsk(dsk_raw),
        "meta": meta,
        "version": version,
        "smart_start": version == VERSION_SMART_START,
        "requested_security_classes": _parse_requested_security_classes(d[9:12]),
    }


def extract_qr_string(text: str) -> str:
    d = _digits_only(text)
    if not d.startswith("90"):
        return ""
    best = ""
    for end in range(90, min(len(d), 200) + 1):
        trial = d[:end]
        if parse_qr_digits(trial):
            best = trial
    if best:
        return best
    parsed = parse_qr_digits(d)
    return parsed["qr"] if parsed else ""


def has_scannable_qr(qr_payload: str | None) -> bool:
    # extract_qr_string only ever returns non-empty when it found a subset that
    # parse_qr_digits fully validated (checksum + DSK) — no extra length threshold
    # needed, and a hardcoded 90 wrongly rejected valid QR codes with a short or
    # missing TLV tail (the spec's actual minimum is 52: see parse_qr_digits).
    return bool(extract_qr_string(qr_payload or ""))


def qr_encode_payload(qr_payload: str | None) -> str | None:
    extracted = extract_qr_string(qr_payload or "")
    return extracted or None


def normalize_fields(manual_code: str = "", qr_payload: str = "") -> dict[str, Any]:
    qr_in = (qr_payload or "").strip()
    qr_extracted = extract_qr_string(qr_in) if qr_in else ""
    parsed = parse_qr_digits(qr_extracted) if qr_extracted else None

    dsk_digits = ""
    if parsed:
        dsk_digits = _digits_only(parsed["dsk"])
    elif is_valid_dsk_formatted(manual_code):
        dsk_digits = _digits_only(manual_code)
    else:
        m = _digits_only(manual_code)
        if len(m) == 40:
            dsk_digits = m

    return {
        "manual_code": format_dsk(dsk_digits) if dsk_digits else (manual_code or "").strip(),
        "qr_payload": parsed["qr"] if parsed else qr_extracted,
        "zwave_pin": pin_from_dsk(dsk_digits or manual_code),
        "zwave_meta": parsed["meta"] if parsed else {},
    }


def code_protocol(data: dict[str, Any]) -> str:
    ct = str(data.get("code_type") or "matter").strip().lower()
    if ct == "zwave":
        return "zwave"
    if ct == "homekit":
        return "homekit"
    if has_scannable_qr(str(data.get("qr_payload") or "")):
        return "zwave"
    qr = str(data.get("qr_payload") or "").strip().upper()
    if qr.startswith("X-HM://"):
        return "homekit"
    return "matter"


def meta_summary(meta: dict[str, Any]) -> str:
    if not meta.get("manufacturer_id"):
        return ""
    parts = [f"Mfg {meta['manufacturer_id']}"]
    if meta.get("product_type") is not None and meta.get("product_id") is not None:
        parts.append(f"Type {meta['product_type']} / ID {meta['product_id']}")
    return " · ".join(parts)

File: anti_matter/app/static/brand/js/zwave-payload.js
/**
 * Z-Wave SmartStart QR (SDS13937) — aligned with zwave-js/qr and Silicon Labs format.
 * QR text: decimal digits only; lead-in 90 ('Z'), version 01 = SmartStart.
 */
(function (global) {
  const LEAD_IN = 90;
  const VERSION_SMART_START = 1;

  // Matches zwave-js's ProvisioningInformationType enum (packages/core/src/qr/definitions.ts).
  const TLV_PRODUCT_TYPE = 0;
  const TLV_PRODUCT_ID = 1;
  const TLV_SUPPORTED_PROTOCOLS = 4;

  function digitsOnly(value) {
    return String(value || "").replace(/\D/g, "");
  }

  /** @returns {string} */
  function formatDsk(dskDigits) {
    const d = digitsOnly(dskDigits);
    if (d.length !== 40) return String(dskDigits || "").trim();
    const parts = [];
    for (let i = 0; i < 8; i++) {
      parts.push(d.slice(i * 5, i * 5 + 5));
    }
    return parts.join("-");
  }

  function parseDskGroups(dskDigits) {
    const d = digitsOnly(dskDigits);
    if (d.length !== 40) return null;
    const groups = [];
    for (let i = 0; i < 8; i++) {
      const g = parseInt(d.slice(i * 5, i * 5 + 5), 10);
      if (Number.isNaN(g) || g < 0 || g > 65535) return null;
      groups.push(g);
    }
    return groups;
  }

  function isValidDskFormatted(value) {
    const s = String(value || "").trim();
    if (!/^\d{5}(-\d{5}){7}$/.test(s)) return false;
    return parseDskGroups(s) !== null;
  }

  function pinFromDsk(dskDigits) {
    const d = digitsOnly(dskDigits);
    return d.length >= 5 ? d.slice(0, 5) : "";
  }

  // Pure-JS SHA-1 (SubtleCrypto's digest() is async-only, but the grid render
  // path — parseQrDigits, called synchronously for every card — needs the
  // checksum decision immediately to know whether to draw a QR <img> tag at
  // all. Verified byte-for-byte against Python's hashlib.sha1 for the exact
  // body strings this file hashes (checksumForBody below).
  function sha1Sync(bytes) {
    let h0 = 0x67452301, h1 = 0xefcdab89, h2 = 0x98badcfe, h3 = 0x10325476, h4 = 0xc3d2e1f0;
    const ml = bytes.length * 8;
    const withOne = new Uint8Array(bytes.length + 1);
    withOne.set(bytes);
    withOne[bytes.length] = 0x80;
    let totalLen = withOne.length;
    while (totalLen % 64 !== 56) totalLen++;
    const padded = new Uint8Array(totalLen + 8);
    padded.set(withOne);
    const dv = new DataView(padded.buffer);
    dv.setUint32(padded.length - 4, ml >>> 0, false);
    dv.setUint32(padded.length - 8, Math.floor(ml / 0x100000000), false);
    for (let chunkStart = 0; chunkStart < padded.length; chunkStart += 64) {
      const w = new Uint32Array(80);
      for (let i = 0; i < 16; i++) {
        w[i] = (padded[chunkStart + i * 4] << 24) | (padded[chunkStart + i * 4 + 1] << 16) |
          (padded[chunkStart + i * 4 + 2] << 8) | padded[chunkStart + i * 4 + 3];
      }
      for (let i = 16; i < 80; i++) {
        const v = w[i - 3] ^ w[i - 8] ^ w[i - 14] ^ w[i - 16];
        w[i] = (v << 1) | (v >>> 31);
      }
      let a = h0, b = h1, c = h2, d = h3, e = h4;
      for (let i = 0; i < 80; i++) {
        let f, k;
        if (i < 20) { f = (b & c) | (~b & d); k = 0x5a827999; }
        else if (i < 40) { f = b ^ c ^ d; k = 0x6ed9eba1; }
        else if (i < 60) { f = (b & c) | (b & d) | (c & d); k = 0x8f1bbcdc; }
        else { f = b ^ c ^ d; k = 0xca62c1d6; }
        const temp = (((a << 5) | (a >>> 27)) + f + e + k + w[i]) >>> 0;
        e = d; d = c; c = ((b << 30) | (b >>> 2)) >>> 0; b = a; a = temp;
      }
      h0 = (h0 + a) >>> 0; h1 = (h1 + b) >>> 0; h2 = (h2 + c) >>> 0; h3 = (h3 + d) >>> 0; h4 = (h4 + e) >>> 0;
    }
    const out = new Uint8Array(20);
    const dv2 = new DataView(out.buffer);
    [h0, h1, h2, h3, h4].forEach((h, i) => dv2.setUint32(i * 4, h, false));
    return out;
  }

  function checksumValid(qrDigits) {
    if (qrDigits.length < 9) return false;
    const given = parseInt(qrDigits.slice(4, 9), 10);
    const body = qrDigits.slice(9);
    const digest = sha1Sync(new TextEncoder().encode(body));
    const expected = (digest[0] << 8) | digest[1];
    return given === expected;
  }

  function parseTlvs(tail) {
    const meta = {};
    let pos = 0;
    while (pos + 4 <= tail.length) {
      const typeCrit = parseInt(tail.slice(pos, pos + 2), 10);
      pos += 2;
      const typ = typeCrit >> 1;
      const len = parseInt(tail.slice(pos, pos + 2), 10);
      pos += 2;
      if (len < 0 || pos + len > tail.length) break;
      const data = tail.slice(pos, pos + len);
      pos += len;
      if (typ === TLV_PRODUCT_TYPE && len >= 10) {
        meta.genericDeviceClass = parseInt(data.slice(0, 5), 10) >> 8;
        meta.specificDeviceClass = parseInt(data.slice(0, 5), 10) & 0xff;
        meta.installerIconType = parseInt(data.slice(5, 10), 10);
      } else if (typ === TLV_PRODUCT_ID && len >= 20) {
        meta.manufacturerId = parseInt(data.slice(0, 5), 10);
        meta.productType = parseInt(data.slice(5, 10), 10);
        meta.productId = parseInt(data.slice(10, 15), 10);
        const app = parseInt(data.slice(15, 20), 10);
        meta.applicationVersion = `${app >> 8}.${app & 0xff}`;
      } else if (typ === TLV_SUPPORTED_PROTOCOLS && (len === 2 || len === 3 || len === 5)) {
        const bits = parseInt(data, 10);
        meta.supportedProtocols = {
          zwave: !!(bits & 0x01),
          zwaveLongRange: !!(bits & 0x02),
        };
      }
    }
    return meta;
  }

  // RequestedKeys byte (3 decimal digits at QR position 9-12) — bit N corresponds to
  // SecurityClass ordinal N in zwave-js: 0=S2_Unauthenticated, 1=S2_Authenticated,
  // 2=S2_AccessControl, 7=S0_Legacy (bits 3-6 are unused/reserved).
  function parseRequestedSecurityClasses(requestedKeysDigits) {
    const bits = parseInt(requestedKeysDigits, 10) || 0;
    return {
      s2Unauthenticated: !!(bits & 0x01),
      s2Authenticated: !!(bits & 0x02),
      s2AccessControl: !!(bits & 0x04),
      s0Legacy: !!(bits & 0x80),
    };
  }

  /**
   * @returns {{ qr: string, dsk: string, pin: string, meta: object, version: number }|null}
   */
  function parseQrDigits(qrDigits) {
    const d = digitsOnly(qrDigits);
    if (d.length < 52 || !d.startsWith("90")) return null;
    const version = parseInt(d.slice(2, 4), 10);
    const dskRaw = d.slice(12, 52);
    if (!parseDskGroups(dskRaw)) return null;
    if (!checksumValid(d)) return null;
    const meta = parseTlvs(d.slice(52));
    return {
      qr: d,
      dsk: formatDsk(dskRaw),
      pin: pinFromDsk(dskRaw),
      meta,
      version,
      smartStart: version === VERSION_SMART_START,
      requestedSecurityClasses: parseRequestedSecurityClasses(d.slice(9, 12)),
    };
  }

  function extractQrString(text) {
    const d = digitsOnly(text);
    if (!d.startsWith("90")) return "";
    let best = "";
    for (let end = 90; end <= Math.min(d.length, 200); end++) {
      const trial = d.slice(0, end);
      if (parseQrDigits(trial)) best = trial;
    }
    if (best) return best;
    const parsed = parseQrDigits(d);
    return parsed ? parsed.qr : "";
  }

  function hasScannableQr(qrPayload) {
    // extractQrString only ever returns non-empty when it found a subset that
    // parseQrDigits fully validated (checksum + DSK) — no extra length threshold
    // needed, and a hardcoded 90 wrongly rejected valid QR codes with a short or
    // missing TLV tail (the spec's actual minimum is 52: see parseQrDigits).
    return !!extractQrString(qrPayload);
  }

  function normalizeFields(manualCode, qrPayload) {
    const qrIn = String(qrPayload || "").trim();
    const qrExtracted = qrIn ? extractQrString(qrIn) : "";
    const parsed = qrExtracted ? parseQrDigits(qrExtracted) : null;

    let dskDigits = "";
    if (parsed) {
      dskDigits = digitsOnly(parsed.dsk);
    } else if (isValidDskFormatted(manualCode)) {
      dskDigits = digitsOnly(manualCode);
    } else {
      const m = digitsOnly(manualCode);
      if (m.length === 40) dskDigits = m;
    }

    return {
      manual_code: dskDigits ? formatDsk(dskDigits) : String(manualCode || "").trim(),
      qr_payload: parsed ? parsed.qr : qrExtracted,
      zwave_pin: dskDigits ? pinFromDsk(dskDigits) : pinFromDsk(manualCode),
      zwave_meta: parsed?.meta || {},
    };
  }

  function codeProtocol(code) {
    const ct = String(code?.code_type || "").toLowerCase();
    if (ct === "zwave") return "zwave";
    if (ct === "homekit") return "homekit";
    if (hasScannableQr(code?.qr_payload)) return "zwave";
    const q = String(code?.qr_payload || "").trim().toUpperCase();
    if (q.startsWith("X-HM://")) return "homekit";
    if (q.startsWith("MT:")) return "matter";
    return "matter";
  }

  function formatDskDisplay(manualCode) {
    const formatted = formatDsk(manualCode);
    return formatted || String(manualCode || "").trim();
  }

  function metaSummary(meta) {
    if (!meta || !meta.manufacturerId) return "";
    const parts = [];
    if (meta.manufacturerId != null) {
      parts.push(`Mfg ${meta.manufacturerId}`);
    }
    if (meta.productType != null && meta.productId != null) {
      parts.push(`Type ${meta.productType} / ID ${meta.productId}`);
    }
    return parts.join(" · ");
  }

  global.AntiMatterZWavePayload = {
    LEAD_IN,
    formatDsk,
    formatDskDisplay,
    pinFromDsk,
    isValidDskFormatted,
    parseQrDigits,
    extractQrString,
    hasScannableQr,
    normalizeFields,
    codeProtocol,
    metaSummary,
    checksumValid,
  };
})(typeof window !== "undefined" ? window : globalThis);
